### DISCUSSION TITLE
supporting projects that are just serverless functions, i.e., no `public/`, just `api/`

### DISCUSSION THREAD

**Discussion by aegatlin** [vercel/vercel#5894] [Ideas]
I've been investigating deploying a pure JSON API to vercel for the past 2 days, and it seems like the serverless functions feature is _not_ intended to be used as a stand-alone feature. That is, where a project is nothing but serverless functions in a top-level `api/` folder.

I can understand why, since it was likely designed in service to Nextjs and other Vercel-hosted SSGs, etc. That said, because you such a simple platform, I would _much rather_ host serverless apis here on Vercel instead of via AWS Lambda or GCP.

You can _essentially_ do this by having an empty `public/index.html` and then only utilizing the `api/` serverless functions to deliver your content. It's a little clunky, especially since the serverless functions don't have a standalone development environment (no `vercel dev` that's doing "just" local serverless function deploys), but I pretty much still prefer it to writing AWS Lambdas.

So, is this on the roadmap? Or, is it a stupid idea for a reason I'm not thinking about? Curious to start a discussion on it :D 

Thanks for reading

**Comment by heysailor** [comment]
You just put your serverless functions in their own folders in API, and use the folders as the routing:


```

// /api/hello.js

module.exports = (req, res) => return 'Hello'

```

After deployment, you can call this at `[yourfunction].vercel.app/api/hello`.

[That's probably what you read in the docs](https://vercel.com/docs/serverless-functions/introduction) - there's nothing to stop you having NO public folder. It's also available with vercel dev locally - you'll need to link to a Vercel project before calling the functions. If there are only serverless functions, that's all vercel dev handles. It's a great dev environment - I leave vercel dev running, and it picks up file changes as I edit, so you get live serverless function responses.

That said, Vercel (when it was Now) used to support even Dockerfile API deployments, and it iterated quickly to what it is now, a service largely oriented towards hosting SSGs, and the API implementations to be hosted elsewhere.

**Reply by aegatlin** [reply]
Hey @heysailor , thanks for the reply. When I tried removing the `public/` folder it wouldn't build on Vercel. I'm curious how you managed to achieve this. For example, if you clone their `custom-build` example repo and remove the `public/` and update the `npm run build` script so that it won't automatically create one, then Vercel will fail to build.

I'll also look more into the `vercel dev` local env more. While it seemed like it _could_ work with a project that was only serverless functions, it didn't seem like it was intended to, or designed to.

Basically, a project that is just a JSON API seems like a second class citizen in Vercel, and I was curious if there was any plans to make it a first class citizen. Even while second class, it can be done. I have successfully deployed my API. But, the experience could be better. Also, if there is no plan to really support this type of usage, then I don't think I want to keep my serverless API hosted here, and would be better served moving it elsewhere.

**Reply by heysailor** [reply]
Hmm I have repos without a public folder - they don't build on the vercel web GUI, but do form a build under the hood; that may be the difference we are seeing. When I use `vercel deploy` on the CLI they progress fine.

You're right though - with the move away from explicit first class support for pure serverless functions, I have an eye towards moving them onto another dedicated host.
